runtime/doc/lsp.txt

```diff
@@ -1323,14 +1323,75 @@ save({lenses}, {bufnr}, {client_id})                 *vim.lsp.codelens.save()*
 ==============================================================================
 Lua module: vim.lsp.semantic_tokens                      *lsp-semantic_tokens*
 
+add_rule({rule}, {opts})                  *vim.lsp.semantic_tokens.add_rule()*
+    Add a new semantic token highlighting rule.
+
+    Example: >lua
+    -- Highlight static local variables
+    vim.lsp.semantic_tokens.add_rule(function(token)
+      if token.modifiers.functionScope and token.modifiers.static then
+        return "MyStaticLocalHL"
+      end
+    end)
+
+    vim.api.nvim_set_hl(0, "MyStaticLocalHL", { fg = "red" })
+<
+
+    Parameters: ~
+      • {rule}  (function) Lua callback invoked on visible tokens. Return a
+                highlight group name to apply that highlight to a token. Args:
+                • token: a table with the fields:
+                  • type: string LSP token type, e.g. 'function' or 'variable'
+                  • modifiers: set of LSP token modifiers, e.g. >lua
+
+                     { defaultLibrary = true, readOnly = true }
+<
+                  • line: line number, 0-based
+                  • start_col: start column, 0-based
+                  • end_col: start column, 0-based
+
+                • bufnr: buffer number
+                • priority: priority assigned to highlight (see below)
+                • namespace: the namespace managed by the semantic token
+                  highlighter. If you apply extmarks inside the rule, you
+                  likely want to use this namespace.
+      • {opts}  (table|nil) Optional keyword arguments:
+                • desc (string|nil): Description of rule, default "none"
+                • bufnr (number|nil): If specified, the rule only applies in
+                  the given buffer.
+                • ft (string|nil): if specified, applies to all files with the
+                  given filetype. Ignored if `bufnr` is specified. If neither
+                  is given, the rule applies globally.
+                • priority (number|nil): Use this priority for highlights
+                  returned by this rule, and supply it as the `priority`
+                  argument to `rule` callbacks. By default, this is 1 higher
+                  than the last rule added with `add_rule`, so rules added
+                  later have higher priority.
+                • pri_stack (number|nil): The number of priority level levels
+                  used by this rule; the default priority of the next added
+                  rule is increased by this amount. Mostly useful if you are
+                  setting extmarks inside a rule. Default 1, or 0 if
+                  `priority` is given.
+
+    Return: ~
+        (number) id of created rule
+
+delete_rule({id})                      *vim.lsp.semantic_tokens.delete_rule()*
```

Comment:
I'm not 100% happy with the semantics here: this function conflates _deleting_ a single rule and _clearing_ all rules.

It would probably be better to make this a
```lua
vim.lsp.semantic_tokens.clear_rules(opts)
```
where `opts` is a table with keys
* `id` (table): id of rule(s) to clear
* `default` (boolean, default to be debated): clear default rules as well

and, possibly and later, `ns` for a namespace.

Reply:
(In any case, common API naming pattern would be `del_rule`, not `delete_rule`.)

Reply:
> I'm not 100% happy with the semantics here: this function conflates _deleting_ a single rule and _clearing_ all rules.

Can you clarify? Are you recommending two functions, `del_rule` and `clear_rules`?

It seems like a common pattern in the api to have `number|nil` arguments, where `number` is the thing (buffer, window, etc) and `nil` is "all the things."

Does an API like this make sense?
```
del_rule({id}, {opt})
```
Parameters:
- {id} (number|table|nil) Id of rule to delete. If table, a list of ids to delete. If nil, delete all rules.
- {opt} (table|nil)
  - keep_default (bool|nil): Default true, don't delete the default when deleting rules

Reply:
> It seems like a common pattern in the api to have number|nil arguments, where number is the thing (buffer, window, etc) and nil is "all the things."

Not really; usually in these instances `nil` means "current <thing>". For situations like these, it's more common to have explicit filters via keywords. 
And we're trying to be stricter about the Lua API design, so I still suggest to move the `id` into `opts`. This won't change the default invocation where `del_rule()` removes all non-default rules but avoids a source of confusion which optional argument is positional and which is keyword-style.

Also, as the default is removing all rules, the name should reflect that, so `clear` instead of `del`.

(But as this is under the LSP umbrella, I defer to @mfussenegger here.)

Reply:
Thank you, this is good feedback! So, the `opts` act as a masking mechanism, and default values indicate "no mask."

This leads to a `clear_rules({opt})`, where `opt` is:
- id (number|table|nil) If defined, only clear rules matching given id(s)
- default (bool|nil) If true, only clear default rules. If false, do not clear default rules.

```lua
clear_rules() -- clear all rules, including default
clear_rules { id = {2, 4} } -- only clear rules 2 and 4
clear_rules { default = true } -- only clear default rule(s)
clear_rules { default = false } -- clear all rules except default rules

-- silly, but it's good to think about interactions
clear_rules { id = 2, default = true } -- clear rule 2 if it's a default rule
clear_rules { id = {1, 2, 3}, default = false } -- clear all of rules 1, 2, 3 that are not default
```

Reply:
>  If true, only clear default rules. If false, do not clear default rules.

Is there a compelling use case for clearing _only_ the default rules? Otherwise I'd avoid the unusual ternary (`true|false|nil`) behavior and just have `default = true` mean "clear **also** default rule(s)".

Reply:
On a related note, is the default rule expected to have stable `id` (at least without deliberate shenanigans)?

Reply:
I imagine the most common use of `clear_rules` would be to clear all rules, with either the intent to entirely disable token highlighting, or to start creating user rules from scratch. For both of these uses, the default rule would be included in "all rules" — it would probably be the only rule!

Because I believe that will be the most common use, I prefer that the no-arguments `clear_rules()` should clear all rules including the default, but it's a weak preference.

I feel like there should always be a way to get back to the default state. That could be:
```lua
st.clear_rules { default = false }
```
That is, opt-out of deleting the default rule, rather than opt-in.

I can't think of any reason why the default rule won't always have the same id; probably `0` or `1`. If we made that part of the API, then users could `clear_rules { id = 1 }`, though I agree it's hard to imagine _why_ that would be useful.

So, it looked like you were suggesting the opt-in:
```lua
clear_rules() -- clear all non-default rules
clear_rules { default = true } -- clear all rules including default
clear_rules { id = {2, 4} } -- clear only rules 2 and 4
```

Where I am suggesting the opt-out:
```lua
clear_rules() -- clear all rules including default
clear_rules { default = false } -- clear all rules excluding default
clear_rules { id = {2, 4} } -- clear only rules 2 and 4
```

If "the default rule has id `1`" is part of the API, then the `default` argument isn't required:
```lua
clear_rules { id = 1 } -- clear default rule
clear_rules() -- clear all rules except default
clear_rules { id = {2, 4} } -- clear only rules 2 and 4
```

Reply:
I have no preference regarding the default for `default`; I just prefer it to be either `true` or `false` rather than `nil` (meaning yet a third thing).

I also assume that "clear only default rule" will not be a very common use case so I wouldn't design the API around it. (Even clearing indvidual rules is kinda pushing it... See above re: don't preemptively implement features without strong personal use cases that you're likely to have to change based on feedback. But I digress.)

So I'm primarily concerned with
1. delete ALL THE RULES
2. delete all rules _except the default static link rule_.